R/kartverket-utils.R

```diff
@@ -0,0 +1,249 @@
+    kommunenavn = mun_name,
+    adressenavn = address_name,
+    adressetekst = address_text,
+    adressetilleggsnavn = address_ext_name,
+    adressekode = address_code,
+    nummer = address_number,
+    bokstav = address_letter,
+    gardsnummer = cadastral_unit_number,
+    bruksnummer = property_unit_number,
+    festenummer = leasehold_number,
+    undernummer = sub_code,
+    bruksenhetsnummer = unit_number,
+    poststed = postal_town,
+    postnummer = postal_code,
+    utkoordsys = crs,
+    treffPerSide = size,
+    side = page,
+    asciiKompatibel = ascii_compatible
+  )
+
+  return(params)
+}
+
+#' Query parameters for Find Address
+#'
+#' Accepted query parameters for Kartverket's endpoint `/punktsok`.
+#'
+#' @inheritParams find_address_from_point
+#' @inheritParams create_params_kv_sok
+#' @keywords internal
+create_params_kv_punktsok <- function(
+    lat, lon,
+    radius,
+    crs,
+    ascii_compatible = TRUE,
+    size = 1000,
+    page = 0
+) {
+
+  # Split CRS
+  crs <- handle_crs(crs)
+
+  params <- list(
+    lat = lat,
+    lon = lon,
+    radius = radius,
+    koordsys = crs$crs_in,
+    utkoordsys = crs$crs_out,
+    treffPerSide = size,
+    side = page,
+    asciiKompatibel = ascii_compatible
+  )
+
+  return(params)
+
+}
+
+#' parse_kartverket
+#' @noRd
+parse_kartverket <- function(resp, parsed, x, radius, crs, ...) {
+  parsed_adresser <- parsed$adresser
+  n_total <- parsed$metadata$totaltAntallTreff
+  n <- nrow(parsed_adresser)
+  if (n_total > n) {
+    msg <- 'Found more addresses than was returned on the first page by the API.'
+    if (rlang::is_interactive()) { #interactive()
+      cli::cli_alert_warning(msg)
+      user_yes <- kv_ask()
+      # if (is.null(user_yes)) cli::cli_alert_danger('A value is needed')
+      if (user_yes) {
+        # n_remaining = n_total - n
+        iter <- ceiling((n_total - n) / n)
+        tmp_resp <- vector('list', iter+1)
+        tmp_parsed <- vector('list', iter+1)
+        tmp_resp[[1]] <- resp
+        tmp_parsed[[1]] <- parsed
+        for (i in seq_len(iter)) {
+          tmp_resp[[i+1]] <-
+            request_kartverket(
+              'punktsok', lat = x[1], lon = x[2], radius = radius,
+              crs = crs, page = i, ...)
+          tmp_parsed[[i+1]] <-
+            parse_response(tmp_resp[[i+1]], simplifyVector = TRUE)
+        }
+        resp <- tmp_resp
+        parsed_adresser <- purrr::map_df(tmp_parsed, function(x) x$adresser)
+      }
+    } else {
+      bullet <- 'Found {n_total} addresses in total, while {n} are currently retrieved.'
+      cli::cli_warn(c(msg, i = bullet))
+    }
+  }
+  list(resp = resp, parsed = parsed_adresser)
+}
+
+
+#' handle_crs
+#' @noRd
+handle_crs <- function(crs) {
+  if (length(crs) == 1) {
+    crs_in <- crs_out <- crs
+  } else if (length(crs) == 2){
+    crs_in <- crs[1]
+    crs_out <- crs[2]
+  } else {
+    cli::cli_abort('`crs` has incorrect length')
+  }
+  list(crs_in = crs_in, crs_out = crs_out)
+}
```

Comment:
Should we handle type errors in the `crs` argument. Both numeric values and character strings will return valid results, but the value must be a valid integer. I.e.:
```r
x <- c(lon = 10.39679, lat = 63.42805)
find_address_from_point(x, crs = 4326) # works
find_address_from_point(x, crs = '4326') # works
find_address_from_point(x, crs = 'EPSG:4326') # fails with API error
```

Reply:
Ah. Yes. Do you think we should handle 'EPSG:4326' for the user or that we just throw an error if the input is incorrect?

Reply:
I'm leaning towards throwing an informative error that `crs` must be coercible to integer.

Reply:
Sounds good.